app/src/main/java/io/pumpkinz/pumpkinreader/service/DataSource.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    public Observable<List<News>> getNew(final int from, final int count, boolean isRefresh) {
+        return listNew(isRefresh)
+                .compose(new NewsTransformer(from, count));
+    }
+
+    private Observable<List<Integer>> listTop(boolean isRefresh) {
+        List<Integer> retval = getNewsFromSp(Constants.TOP_FILE_SP, Constants.TOP_VAL_SP, ctx);
+
+        if (retval.isEmpty() || isRefresh) {
+            return RestClient.service().listTop()
+                    .doOnNext(new putToSpAction(ctx, Constants.TOP_FILE_SP, Constants.TOP_VAL_SP));
+        } else {
+            return Observable.just(retval);
+        }
+    }
+
+    private Observable<List<Integer>> listNew(boolean isRefresh) {
+        List<Integer> retval = getNewsFromSp(Constants.NEW_FILE_SP, Constants.NEW_VAL_SP, ctx);
+
+        if (retval.isEmpty() || isRefresh) {
+            return RestClient.service().listNew()
+                    .doOnNext(new putToSpAction(ctx, Constants.NEW_FILE_SP, Constants.NEW_VAL_SP));
+        } else {
+            return Observable.just(retval);
+        }
+    }
+
+    //TODO: add listAsk, listJob, and listShow
+
+    private List<Integer> getNewsFromSp(String newsFileKey, String newsValKey, Context context) {
+        List<Integer> retval = new ArrayList<>();
+
+        SharedPreferences topStoriesSp = context.getSharedPreferences(
+                newsFileKey, Context.MODE_PRIVATE);
+        String topStories = topStoriesSp.getString(newsValKey, "");
+
+        if (!topStories.isEmpty()) {
+            retval = Util.split(topStories, "|");
+        }
+
+        return retval;
+    }
+
+    private class putToSpAction implements Action1<List<Integer>> {
+
+        private Context context;
+        private String SP_FILE_KEY;
+        private String SP_VAL_KEY;
+
+        public putToSpAction(Context context, String SP_FILE_KEY, String SP_VAL_KEY) {
+            this.context = context;
+            this.SP_FILE_KEY = SP_FILE_KEY;
+            this.SP_VAL_KEY = SP_VAL_KEY;
+        }
+
+        @Override
+        public void call(List<Integer> integers) {
+            String input = Util.join(integers, '|');
+            SharedPreferences topStoriesSp = context.getSharedPreferences(
+                    SP_FILE_KEY, Context.MODE_PRIVATE);
+            SharedPreferences.Editor editor = topStoriesSp.edit();
+
+            editor.putString(SP_VAL_KEY, input);
+            editor.commit();
+        }
+
+    }
+
+    private class NewsTransformer implements Observable.Transformer<List<Integer>, List<News>> {
+
+        final List<Integer> topNewsIds = new ArrayList<>();
+        final Dictionary<Integer, News> dict = new Hashtable<>();
+        private int from;
+        private int count;
+
+        public NewsTransformer(int from, int count) {
+            this.from = from;
+            this.count = count;
+        }
+
+        @Override
+        public Observable<List<News>> call(Observable<List<Integer>> integers) {
```

Comment:
The parameter name is really not explaining anything about what the parameter is, aside of the type of it.
Would be better to name it ```ids```